metron-platform/metron-enrichment/Performance.md

```diff
@@ -0,0 +1,527 @@
+
+### Prepare Enrichment Data
+
+The Alexa Top 1 Million was used as an data source for these benchmarks.
+
+1. Download the [Alexa Top 1 Million](http://s3.amazonaws.com/alexa-static/top-1m.csv.zip).
+
+2. For each hostname, query DNS to retrieve an associated IP address.  
+
+	A script like the following can be used for this.  There is no need to do this for all 1 million entries in the data set. Doing this for around 10,000 records is sufficient.
+
+	```python
+	import dns.resolver
+	import csv
+
+	resolver = dns.resolver.Resolver()
+	resolver.nameservers = ['8.8.8.8', '8.8.4.4']
+
+	with open('top-1m.csv', 'r') as infile:
+	  with open('top-1m-with-ip.csv', 'w') as outfile:
+
+	    reader = csv.reader(infile, delimiter=',')
+	    writer = csv.writer(outfile, delimiter=',')
+	    for row in reader:
+
+	      host = row[1]
+	      try:
+	        response = resolver.query(host, "A")
+	        for record in response:
+	          ip = record
+	          writer.writerow([host, ip])
+	          print "host={}, ip={}".format(host, ip)
+
+	      except:
+	        pass
+	```
+
+3. The resulting data set contains an IP to hostname mapping.
+	```bash
+	$ head top-1m-with-ip.csv
+	google.com,172.217.9.46
+	youtube.com,172.217.4.78
+	facebook.com,157.240.18.35
+	baidu.com,220.181.57.216
+	baidu.com,111.13.101.208
+	baidu.com,123.125.114.144
+	wikipedia.org,208.80.154.224
+	yahoo.com,98.139.180.180
+	yahoo.com,206.190.39.42
+	reddit.com,151.101.1.140
+	```
+
+### Load HBase with Enrichment Data
+
+1. Create an HBase table for this data.  
+
+	Ensure that the table is evenly distributed across the HBase nodes.  This can be done by pre-splitting the table or splitting the data after loading it.  
+
+	```
+	create 'top-1m', 't', {SPLITS => ['2','4','6','8','a','c','e']}
+	```
+
+1. Create a configuration file called `extractor.json`.  This defines how the data will be loaded into the HBase table.
+
+	```bash
+	> cat extractor.json
+	{
+	    "config": {
+	        "columns": {
+	            "host" : 0,
+	            "ip": 1
+	        },
+	        "indicator_column": "ip",
+	        "type": "top-1m",
+	        "separator": ","
+	    },
+	    "extractor": "CSV"
+	}
+	```
+
+1. Use the `flatfile_loader.sh` to load the data into the HBase table.
+	```
+	$METRON_HOME/bin/flatfile_loader.sh \
+		-e extractor.json \
+		-t top-1m \
+		-c t \
+		-i top-1m-with-ip.csv
+	```
+
+### Configure the Enrichments
+
+1. Define the Enrichments using the REPL.
+
+	```
+	> $METRON_HOME/bin/stellar -z $ZOOKEEPER
+	Stellar, Go!
+
+	[Stellar]>>> conf
+	{
+	  "enrichment": {
+	    "fieldMap": {
+	     "stellar" : {
+	       "config" : {
+	         "geo" : "GEO_GET(ip_dst_addr)",
+	         "top1m" : "ENRICHMENT_GET('top-1m', ip_dst_addr, 'top-1m', 't')",
+	         "local" : "IN_SUBNET(ip_dst_addr, '192.168.0.0/24')"
+	       }
+	     }
+	    },
+	    "fieldToTypeMap": {
+	    }
+	  },
+	  "threatIntel": {
+	  }
+	}
+	[Stellar]>>> CONFIG_PUT("ENRICHMENT", conf, "asa")
+	```
```

Comment:
If you have a blank line in a codeblock nested under a list it doesn't format properly in the site-book.  An easy fix is to remove the empty newline, thoughts?

Reply:
Everything should be addressed @JonZeolla .  Thanks